### PR TITLE
Fix most package error from ansi-tests + new regression-tests

### PR DESCRIPTION
**Pull request by kpoeck**
* now only these 3 remains unsolved `feature-assorted-package-fixes
* solves #1141

### CODE REVIEW

**Comment by Bike** [comment]
By coincidence I have also been working on the package test failures, though I was focusing more on name conflicts than on package creation/deletion. I think some of your correctable errors here aren't totally correct - the CLHS is pretty specific about what restarts need to be able to do (11.1.1.2.5), but the ansi-tests can't cover that because the names of the restarts aren't standard. Many of those actions are things that would be pretty annoying to write in C++. For example, for `import`, "A name conflict in import with a symbol already present in the package may be resolved by uninterning that symbol, or by not doing the import."; since there are two possible actions this can't be done with just `cerror`.

Long story short, what I'm going to do is merge your changes into my working branch, then merge that branch. I'll discard some of your changes but keep the package designator stuff and the new tests.

Also, what's with this bit? https://github.com/clasp-developers/clasp/pull/1152/commits/3a6e0c85b9b6f9560f0c190efbaab521bc4cfdf1#diff-e83b41979710f43a5ef9eda0b8c6122b868de51904efa7423f2ebb06e571f6b7R902-R904 CLHS says in 11.1.1.2.3 that `import` does not change the home package of the importees.

**Comment by kpoeck** [comment]
regarding the home package of uninterned symbols, clhs does say it.

http://www.lispworks.com/documentation/HyperSpec/Body/f_import.htm

"If any symbol to be imported has no home package (i.e., (symbol-package symbol) => nil), import sets the home package of the symbol to package."

Regarding the correctable errors I understand clhs differently, but Ithan  am no native speaker. I understand "may" as a possibility, not as a rule.

**Comment by Bike** [comment]
ah, you're right about import. cool, the CLHS contradicts itself. setting the home package seems fine then.

**Comment by Bike** [comment]
Does this work in your branch? It doesn't in my working branch that I've merged your changes into, and I'm not sure if it's something I did.

```lisp
(let ((par0 (make-package "PAR0")) (par1 (make-package "PAR1")) (par2 (make-package "PAR2")) (chil (make-package "CHIL")))
  (use-package (list par0 par1 par2) chil)
  (delete-package chil) (delete-package par0) (delete-package par1) (delete-package par2))
```

For me, the used-by lists are getting messed up somehow, so par0 can't be deleted because chil depends on it, even though chil has already been deleted.

**Comment by kpoeck** [comment]
I get in my branch
```lisp
Condition of type: SIMPLE-PACKAGE-ERROR
Package with designator #<PACKAGE PAR0> is used by at least one other package:
Available restarts:
(use :r1 to invoke restart 1, etc.)

1. (CONTINUE) Remove dependency in other packages
2. (RESTART-TOPLEVEL) Go back to Top-Level REPL.
````

**Comment by kpoeck** [comment]
there might be an error even before my changes
```lisp
(defvar *par0*)
(defvar *par1*)
(defvar *par2*)
(defvar *chil*)

(setq *par0* (make-package "PAR0" :use nil)
      *par1* (make-package "PAR1" :use nil)
      *par2* (make-package "PAR2" :use nil)
      *chil* (make-package "CHIL11" :use nil))

(use-package (list *par0* *par1* *par2*) *chil*)
(delete-package *chil*)

(package-used-by-list *par0*) -> nil
(package-used-by-list *par1*) -> (#<PACKAGE >)
(package-used-by-list *par2*) -> nil
````
and unsurprisingly `(delete-package *par1*)` gives the cerror.

So my first suspicion is, the error was there before my changes, but now it is noted.

**Comment by kpoeck** [comment]
I did put the following in Package_O::unusePackage_no_outer_lock:
```c++
printf("unuse Package %s from Package %s\n",
         usePackage->_Name->get_std_string().c_str(),
                 me->_Name->get_std_string().c_str()); 
````
this prints:for (delete-package *chil*)
unuse Package PAR0 from Package CHIL11
unuse Package PAR2 from Package CHIL11
unuse Package PAR2 from Package CHIL11

So in `Package_O::unusePackage_no_outer_lock` we do `this->_UsingPackages.erase(it);`
while in `cl__delete_package` we do 
```c++
 for (auto pi : pkg->_UsingPackages) {
    if (pi.notnilp())
      pkg->unusePackage_no_outer_lock(pi);
  }
`````

do we modify _UsingPackages while iterating over it, does not look like a great idea

**Comment by kpoeck** [comment]
ok, changed the code in cl__delete_package to:
```c++
// need to grab before the readwriteLock
  List_sp usedList = pkg->packageUseList();
  List_sp usedByList = pkg->packageUsedByList();
  
  WITH_PACKAGE_READ_WRITE_LOCK(pkg);
  for (auto pi : usedList) {
    Package_sp piLocal = gc::As<Package_sp>(CONS_CAR(pi));
    pkg->unusePackage_no_outer_lock(piLocal);
  }

  for (auto pi : usedByList) {
    Package_sp piLocal = gc::As<Package_sp>(CONS_CAR(pi));
    piLocal->unusePackage_no_inner_lock(pkg);
  }
`````
and your example works now fine.

Will update the pr with the change and a regression-test

**Comment by kpoeck** [comment]
updated the pr as commented

**Comment by Bike** [comment]
yeah, probably the use/used-by lists were always screwed up, but we just ignored it in `delete-package`.

**Comment by Bike** [comment]
alright, i've merged this into my branch and all the waf tests pass now. after i run ansi tests again i'll merge into master
